bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java

```diff
@@ -70,30 +75,42 @@ public synchronized void returnToPool() {
     }
   }
 
-  private final List<Channel> channels;
+  private final AtomicReference<ImmutableList<Channel>> channels = new AtomicReference<>();
   private final AtomicInteger requestCount = new AtomicInteger();
-  private final List<HeaderInterceptor> headerInterceptors;
+  private final ImmutableList<HeaderInterceptor> headerInterceptors;
   private final AtomicInteger totalSize;
   private final ChannelFactory factory;
   private int reservedChannelCount = 0;
+  private final String authority;
 
   public ChannelPool(List<HeaderInterceptor> headerInterceptors, ChannelFactory factory)
       throws IOException {
-    this.channels = new ArrayList<>();
-    channels.add(factory.create());
+    Channel channel = factory.create();
+    this.channels.set(ImmutableList.of(channel));
+    authority = channel.authority();
     totalSize = new AtomicInteger(1);
     this.factory = factory;
-    this.headerInterceptors = headerInterceptors;
+    if (headerInterceptors == null) {
+      this.headerInterceptors = ImmutableList.of();
+    } else {
+      this.headerInterceptors = ImmutableList.copyOf(headerInterceptors);
+    }
   }
 
   public void ensureChannelCount(int capacity) throws IOException {
     if (totalSize.get() < capacity) {
-      synchronized (this) {
-        int unreservedCapacity = capacity - reservedChannelCount;
-        for (int i = channels.size(); i < unreservedCapacity; i++) {
-          channels.add(factory.create());
+      synchronized(this){
+        if (totalSize.get() < capacity) {
+          ImmutableList.Builder<Channel> channelsBuilder = ImmutableList.builder();
+          List<Channel> current = channels.get();
+          channelsBuilder.addAll(current);
+          int unreservedCapacity = capacity - reservedChannelCount;
+          for (int i = current.size(); i < unreservedCapacity; i++) {
+            channelsBuilder.add(factory.create());
+          }
+          channels.set(channelsBuilder.build());
```

Comment:
Building an immutableList is definitely slower than modifying a List.  I'm not concerned about the difference, though, since this happens infrequently.  The point is to make access of the List consistent for the getNextChannel() method which gets called a few orders of magnitude more often than any of the methods that build a new ImmutableList.  This change allowed me to remove the synchronization from getNextChannel() since we're now guaranteed to have a consistent list to get a size on and then do a get() on.

ensureChannelCount's expansion logic should happen once or twice for an entire application.  reserveChannel() may be called infrequently for scans.  getNextChannel() gets called for every operation.